src/api/key_store.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use alloc::vec::Vec;
+use crypto::ecdsa::SecKey;
+use persistent_store::StoreError;
+use rng256::Rng256;
+
+use crate::env::Env;
+
+/// Provides storage for secret keys.
+///
+/// Implementations may use the environment store: [`STORE_KEY`] is reserved for this usage.
+pub trait KeyStore {
+    /// Returns the AES key for key handles encryption.
+    fn key_handle_encryption(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Returns the key for key handles authentication.
+    fn key_handle_authentication(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Derives an ECDSA private key from a seed.
+    ///
+    /// The result is big-endian.
+    fn derive_ecdsa(&mut self, seed: &[u8; 32]) -> Result<[u8; 32], Error>;
+
+    /// Generates a seed to derive an ECDSA private key.
+    fn generate_ecdsa_seed(&mut self) -> Result<[u8; 32], Error>;
+
+    /// Resets the key store.
+    fn reset(&mut self) -> Result<(), Error>;
+}
+
+/// Key store errors.
+///
+/// They are deliberately indistinguishable to avoid leaking information.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct Error;
+
+/// Key of the environment store reserved for the key store.
+pub const STORE_KEY: usize = 2046;
+
+/// Implements a default key store using the environment rng and store.
+pub trait Helper: Env {}
```

Comment:
Is this generally how we intend to implement common behavior between different `Env`s? Basically a mor eexplicit default next to the definition of the trait?
What if we eventually want 2 of these helpers? Can't we give it a better name, like `StorageKeyStore` (not sure if that actually is a good name though)?

Reply:
> Is this generally how we intend to implement common behavior between different Envs?

I think that's one way to provide helpers. Another would be to just provide simpler functions to compose. I think this is to be decided on a case by case basis.

> What if we eventually want 2 of these helpers?

We would have 2 helpers. But until it happens we have no idea how they would differ, so I would not try to over-specify.

Reply:
Since it's public, I thought a more telling name would be appropriate, but no super strong opinion, so your call. Approved.